### DISCUSSION TITLE
CUSBKeyboardDevice does not expose the CKeyboardBehaviour class or functions

### DISCUSSION THREAD

**Discussion by ArnoldUK** [rsta2/circle#364] [Q&A] [answered]
I'm playing around with the `08-usbkeyboard` sample which is working fine. I was also wanting to use the features of the `CKeyboardBehaviour` class as it has more capable and useful functions for handling the keyboard. In particular the `KeyPressed`, `KeyReleased` and `GenerateEvent` functions.

If the callback functions for the Key handlers set by the `CUSBKeyboardDevice` which are then registered with the `CKeyboardBehaviour` class, why is it not possible to access the `CKeyboardBehaviour` interface from the `CUSBKeyboardDevice` instance ? What is the correct way to use an instance of the `CKeyboardBehaviour` class ?

The `CKeyboardBehaviour` class does not have a constructor to pass the `CUSBHIDDevice`.
So I can't use code like this:
```
CUSBKeyboardDevice  * volatile pKeyboard;
pKeyboard = (CUSBKeyboardDevice *)pDeviceNameService->GetDevice("ukbd1", FALSE);
if (pKeyboard == 0) return false;
```
If the `CUSBKeyboardDevice` exposed the `CKeyboardBehaviour m_Behaviour;` member as public and static then it would make more sense to me to have access to those class functions, instead of creating another instance of  the `CKeyboardBehaviour` class.

Thanks.

**Comment by rsta2** [comment]
The [USB keyboard driver](https://circle-rpi.readthedocs.io/en/45.0/devices/character-devices.html#cusbkeyboarddevice) supports a cooked mode with escape sequences and a raw mode, which provides the raw USB keyboard codes and modifier key info. There is no way to access the `CKeyboardBehaviour` member of the driver directly, because it is intended for internal implementation purpose only.

**Comment by ArnoldUK** [comment]
I see how the USB Keyboard driver class works, but I'm not quite sure why the `CKeyboardBehaviour` has the redundant `KeyPressed`, `KeyReleased` and `GenerateEvent`. I cannot find anywhere within the Circle SDK where these functions are used by any other class, not even within the `CUSBKeyboardDevice` class. Also the `CUSBKeyboardDevice::ReportHandler` function appears to be redundant.

However, I do see the `CKeyboardBehaviour` class as a useful template to add to a project and use it alongside the `CUSBKeyboardDevice` driver like so:
```
CUSBKeyboardDevice  * volatile pKeyboard;
pKeyboard = (CUSBKeyboardDevice *)pDeviceNameService->GetDevice("ukbd1", FALSE);
if (pKeyboard == 0) return false;

pKeyboard->Configure(); //DOES THIS NEED TO BE CALLED ???

pKeyboard->RegisterRemovedHandler(MyKeyboardRemovedHandler);
pKeyboard->RegisterKeyStatusHandlerRaw(MyKeyStatusHandlerRaw);
pKeyboard->RegisterShutdownHandler(MyShutdownHandler);
pKeyboard->RegisterKeyPressedHandler(MyKeyPressedHandler);

CKeyboardBehaviour m_Behaviour;
m_Behaviour.RegisterKeyPressedHandler (MyKeyPressedHandler);
m_Behaviour.RegisterShutdownHandler (MyShutdownHandler);
m_Behaviour.RegisterSelectConsoleHandler (MySelectConsoleHandler);

//I can then use the CKeyboardBehaviour function calls which are not exposed in CUSBKeyboardDevice:
m_Behaviour.GenerateKeyEvent(ucKeyCode);
m_Behaviour.KeyPressed(ucKeyCode);
m_Behaviour.KeyReleased(ucKeyCode);
```

My point was that it seems unnecessary creating another instance of `CKeyboardBehaviour` when this class is created and used within the `CUSBKeyboardDevice` but is private and inaccessible. I'm not an expert here, but I'm sure you can correct me.

**Reply by rsta2** [reply]
Believe me, the class `CKeyboardBehaviour` will not help you. It implements the functions, needed internally by the USB keyboard driver in cooked mode. You could use the raw keyboard mode and use `CKeyboardBehaviour` to implement the cooked mode again by yourself. But why should you do this, because it is already provided by the USB keyboard driver (class `CUSBKeyboardDevice`).

*sample/08-keyboard* shows, how the keyboard cooked mode and alternatively the raw mode will be activated and used. `CUSBKeyboardDevice::Configure()` is called internally in the USB stack and must not be called by the application. Please also have a look into the [docs](https://circle-rpi.readthedocs.io/en/45.0/devices/character-devices.html#cusbkeyboarddevice).

BTW. These methods are not redundant. They are called by `CUSBKeyboardDevice` (`KeyPressed()`, `KeyReleased()`, inside `CKeyboardBehaviour` itself (`GenerateEvent()`) or by the base class `CUSBHIDDevice` (`ReportHandler()`).

**Comment by ArnoldUK** [comment]
I was wanting to use the `CKeyboardBehaviour.KeyReleased` function but that is not exposed by the `CUSBKeyboardDevice`.
Only registers for KeyPressed are exposed for registering your own handler to these functions. 

How is it possible to register a KeyReleased handler in a similar fashion to KeyPressed ?  I know I can add my own code for detecting when a key was last pressed and then released. But I thought it would useful for the `CUSBKeyboardDevice` class to also have a KeyReleased event handler. I hope this makes more sense.

**Reply by rsta2** [reply]
OK, if you need the key-released info, you have to use the raw mode of the USB keyboard driver. I think, this is, what most emulator developers do. Perhaps you can re-use some code from the Circle keyboard classes for your purpose, but you have to modify it.

**Comment by ArnoldUK** [comment]
Thanks @rsta2 that is what I have ended up doing and the raw handler sends 0 for parameters when all keys are released.
Only issue now is a possible bug with the key repeat functions. I have set the `#define REPEAT_ENABLE` in the `keyboardbehaviour.cpp` file and re-compiled circle but the key repeat does not work. I'm usung the raw capture handler. If I set the `mixedmode` to TRUE then my kernel freezes halfway through booting when initializing the USB keyboard device. There are no error messages either. `mixedmode` set to FALSE works fine but without key repeat functionality.

I will upload a sample of my keyboard testing kernel, maybe you can spot something out of place.
Appreciate your time and patience.

**Reply by rsta2** [reply, marked as answer]
You are welcome. Please note that the keyboard repeat function only works in cooked mode. There is no repeat in raw mode. There have also been issues with keyboard repeat. Sometimes the repeat did not stop. That's why it is not enabled by default. But I've tested it here again with *sample/08-usbkeyboard* (also in mixed mode) on a RPi 4 and it worked all right.

**Comment by ArnoldUK** [comment]
Ok,  I'm using a Rpi 1 Model B for build and test. I'm sure I could easily code my own key repeat methods. I think some of my issues may have been casued by not having `USE_USB_SOF_INTR ` set in sysconfig.h. 
Cheers for the tips.

**Comment by ArnoldUK** [comment]
I've managed to sort out all the keyboard issues by changing the mixedmode boolean parameter in `functionpKeyboard->RegisterKeyStatusHandlerRaw(m_pKeyStatusHandlerRaw, TRUE);` to TRUE.

Key repeats work perfectly now in the cooked handler and I can also capture the modifer keys in the raw mode handler.
I should read the docs more often as the mixed mode operation was explained in the docs, I didn't understand the modes fully but now I have a much better understanding.
